File: sling-react-core/src/main/java/com/cathcart93/sling/models/AccordionController.java
package com.cathcart93.sling.models;

import com.cathcart93.sling.BaseController;
import com.google.gson.annotations.Expose;
import org.apache.sling.api.resource.Resource;
import org.apache.sling.models.annotations.Model;

import javax.annotation.PostConstruct;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Model(adaptables = Resource.class)
public class AccordionController extends BaseController{
    @Expose
    private String __type = "Accordion";

    @Expose
    private List<AccordionItemModel> items;

    @Override
    @PostConstruct
    public void init(){
        items = StreamSupport.stream(resource.getChildren().spliterator(), false)
                .map(item -> {
                    AccordionItemModel itemModel = item.adaptTo(AccordionItemModel.class);
                    itemModel.init();
                    return itemModel;
                })
                .collect(Collectors.toList());
    }
}

File: sling-react-components/src/main/kotlin/com/cathcart93/sling/components/models/AccordionController.kt
package com.cathcart93.sling.components.models

import com.cathcart93.sling.core.IReactController
import com.cathcart93.sling.core.ReactController
import com.cathcart93.sling.core.ReactProp
import org.apache.sling.api.resource.Resource
import org.apache.sling.models.annotations.Model
import org.apache.sling.models.annotations.injectorspecific.ChildResource
import org.apache.sling.models.annotations.injectorspecific.SlingObject

import javax.annotation.PostConstruct

@Model(adaptables = [Resource::class])
@ReactController(componentName = "Accordion")
class AccordionController : IReactController {

    @ReactProp
    private lateinit var  items: List<AccordionItemModel>

    @SlingObject
    private lateinit var resource: Resource

    @PostConstruct
    fun init() {
        items = resource.children.mapNotNull {
            it.adaptTo(AccordionItemModel::class.java)
        }
    }
}
